src/Lecture4.hs

```diff
@@ -134,8 +137,24 @@ errors. We will simply return an optional result here.
 🕯 HINT: Use the 'readMaybe' function from the 'Text.Read' module.
 -}
 
+split :: String -> [String]
+split [] = []
+split l =
+  h : split (drop 1 t)
+  where
+    (h, t) = break (== ',') l
+
+parseCost :: String -> Maybe Int
+parseCost c = do
+  n <- readMaybe c
+  guard $ n >= 0
+  pure n
+
 parseRow :: String -> Maybe Row
-parseRow = error "TODO"
+parseRow l = do
+  [prod, trade, cost] <- Just $ split l
+  guard $ prod /= ""
+  readMaybe trade >>= \t -> parseCost cost >>= \c -> pure $ Row prod t c
```

Comment:
Awesome implementation using the `MonadFail` instance for `Maybe` and `do`-notation? 👏🏻